# 1. Pipeline Function
The pipeline function returns an end to end object that performs an NLP task on one or several texts.

### Text Classification

In [ ]:
from transformers import pipeline

pipe = pipeline("text-classification")
pipe("This restaurant is awesome")

No model was supplied, defaulted to distilbert/distilbert-base-uncased-finetuned-sst-2-english and revision 714eb0f (https://huggingface.co/distilbert/distilbert-base-uncased-finetuned-sst-2-english).
Using a pipeline without specifying a model name and revision in production is not recommended.
/usr/local/lib/python3.12/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


config.json:   0%|          | 0.00/629 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/268M [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/48.0 [00:00<?, ?B/s]

vocab.txt: 0.00B [00:00, ?B/s]

Device set to use cuda:0


[{'label': 'POSITIVE', 'score': 0.9998743534088135}]

##### If you want to use a specific model from the hub you can ignore the task if the model on the hub already defines it.

In [ ]:
pipe = pipeline(model="FacebookAI/roberta-large-mnli")
pipe("This restraurant is awesome")

config.json:   0%|          | 0.00/688 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/1.43G [00:00<?, ?B/s]

Some weights of the model checkpoint at FacebookAI/roberta-large-mnli were not used when initializing RobertaForSequenceClassification: ['roberta.pooler.dense.bias', 'roberta.pooler.dense.weight']
- This IS expected if you are initializing RobertaForSequenceClassification from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing RobertaForSequenceClassification from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).


tokenizer_config.json:   0%|          | 0.00/25.0 [00:00<?, ?B/s]

vocab.json: 0.00B [00:00, ?B/s]

merges.txt: 0.00B [00:00, ?B/s]

tokenizer.json: 0.00B [00:00, ?B/s]

Device set to use cuda:0


[{'label': 'NEUTRAL', 'score': 0.6371732950210571}]

##### To call a pipeline on many items, you can call it with a list.

In [ ]:
pipe = pipeline("text-classification")
pipe(['This restraunt is awesome', "This restruant is awful"])

No model was supplied, defaulted to distilbert/distilbert-base-uncased-finetuned-sst-2-english and revision 714eb0f (https://huggingface.co/distilbert/distilbert-base-uncased-finetuned-sst-2-english).
Using a pipeline without specifying a model name and revision in production is not recommended.
Device set to use cuda:0


[{'label': 'POSITIVE', 'score': 0.9998738765716553},
 {'label': 'NEGATIVE', 'score': 0.9997547268867493}]

### Pipeline + Dataset

In [ ]:
from transformers import pipeline

# Create a sentiment analyzer
analyzer = pipeline("sentiment-analysis")

# Test with simple examples
texts = [
    "I love this product! It's amazing!",
    "This is the worst thing ever.",
    "It's okay, nothing special."
]

for text in texts:
    result = analyzer(text)[0]
    emotion = "😊" if result['label'] == 'POSITIVE' else '😞'
    print(f"{emotion} '{text}' -> {result['label']} ({result['score']:.2%})")

No model was supplied, defaulted to distilbert/distilbert-base-uncased-finetuned-sst-2-english and revision 714eb0f (https://huggingface.co/distilbert/distilbert-base-uncased-finetuned-sst-2-english).
Using a pipeline without specifying a model name and revision in production is not recommended.
Device set to use cuda:0


😊 'I love this product! It's amazing!' -> POSITIVE (99.99%)
😞 'This is the worst thing ever.' -> NEGATIVE (99.97%)
😞 'It's okay, nothing special.' -> NEGATIVE (81.90%)


### 1.1 Sentiment Analysis

In [ ]:
from transformers import pipeline

classifier = pipeline("sentiment-analysis")
classifier("I've been waiting for a HuggingFace course.")

No model was supplied, defaulted to distilbert/distilbert-base-uncased-finetuned-sst-2-english and revision 714eb0f (https://huggingface.co/distilbert/distilbert-base-uncased-finetuned-sst-2-english).
Using a pipeline without specifying a model name and revision in production is not recommended.
Device set to use cuda:0


[{'label': 'NEGATIVE', 'score': 0.9645649194717407}]

#### What’s happening under the hood:
- Downloads and loads distilbert-base-uncased-finetuned-sst-2-english
- Tokenizes input with AutoTokenizer
- Runs model inference with AutoModelForSequenceClassification
- Applies softmax to logits
- Returns class label and probability

### 1.2 Using a Custom Model

In [ ]:
classifier = pipeline(
    "sentiment-analysis",
    model="nlptown/bert-base-multilingual-uncased-sentiment"
)
print(classifier("C'est incroyable!"))


config.json:   0%|          | 0.00/953 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/669M [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/39.0 [00:00<?, ?B/s]

vocab.txt: 0.00B [00:00, ?B/s]

special_tokens_map.json:   0%|          | 0.00/112 [00:00<?, ?B/s]

Device set to use cuda:0


[{'label': '5 stars', 'score': 0.8434332013130188}]


### 1.3 Text Generation Pipeline
Will complete the line automatically

In [ ]:
generator = pipeline("text-generation")
generator("In this course, we'll teach you how to ")

No model was supplied, defaulted to openai-community/gpt2 and revision 607a30d (https://huggingface.co/openai-community/gpt2).
Using a pipeline without specifying a model name and revision in production is not recommended.


config.json:   0%|          | 0.00/665 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/548M [00:00<?, ?B/s]

generation_config.json:   0%|          | 0.00/124 [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/26.0 [00:00<?, ?B/s]

vocab.json: 0.00B [00:00, ?B/s]

merges.txt: 0.00B [00:00, ?B/s]

tokenizer.json: 0.00B [00:00, ?B/s]

Device set to use cuda:0
Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.


[{'generated_text': "In this course, we'll teach you how to \xa0create a unique, customisable and powerful toolkit for your own applications and applications, with all the tools you need to build your own automation.\nI'm happy to announce that all of my courses will be available at my website after the end of May. I'll be giving away three free courses at the end of the year, but I will not be doing them one after the other. I've already seen a few people wanting to sign up for my courses, so I'll be providing another one at the end of the year.\nI've also put together a list of more than 20 of my favourite courses to check out.\nIf you're interested in getting an email about my course, you can follow me on\xa0 Twitter \xa0or\xa0 Facebook. I'll be posting my course-lists and updates on my\xa0 blog, so if you'd like to subscribe, you can follow me on\xa0 Facebook \xa0and\xa0 Instagram.\nThe course descriptions and videos are given in the order that they appear.\nCourse Description\nI'l

We can also specify the specific model, the maximum length of output (words) and number of sequences i.e., unique outputs.

In [ ]:
generator = pipeline("text-generation",model="distilgpt2")
generator(
    "In this course, we will teach you how to",
    max_length=30,
    num_return_sequences=2
    )

config.json:   0%|          | 0.00/762 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/353M [00:00<?, ?B/s]

generation_config.json:   0%|          | 0.00/124 [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/26.0 [00:00<?, ?B/s]

vocab.json:   0%|          | 0.00/1.04M [00:00<?, ?B/s]

merges.txt:   0%|          | 0.00/456k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/1.36M [00:00<?, ?B/s]

Device set to use cuda:0
Truncation was not explicitly activated but `max_length` is provided a specific value, please use `truncation=True` to explicitly truncate examples to max length. Defaulting to 'longest_first' truncation strategy. If you encode pairs of sequences (GLUE-style) with the tokenizer you can select this strategy more precisely by providing a specific strategy to `truncation`.
Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.
Both `max_new_tokens` (=256) and `max_length`(=30) seem to have been set. `max_new_tokens` will take precedence. Please refer to the documentation for more information. (https://huggingface.co/docs/transformers/main/en/main_classes/text_generation)


[{'generated_text': 'In this course, we will teach you how to practice yoga, how to become an expert in life, and how to become an expert in life. We will teach you how to practice yoga, how to become an expert in life, and how to become an expert in life.\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n'},
 {'generated_text': 'In this course, we will teach you how to use the Java JVM to build Java applications in a more flexible way.\n\n\n\n\nThe Java JVM (JVM) is a multi-platform, high-performance, multi-platform, multi-threaded, multi-threaded, multi-core Java application.\nIn order to learn more about the Java JVM, pl

### 1.4 Fill Mask Pipeline
The fill mask pipeline will predict missing words in a sentence.

In [ ]:
unmasker = pipeline("fill-mask")
unmasker(
    "This couse will teach you all about <mask> models.", top_k=2
)

No model was supplied, defaulted to distilbert/distilroberta-base and revision fb53ab8 (https://huggingface.co/distilbert/distilroberta-base).
Using a pipeline without specifying a model name and revision in production is not recommended.


config.json:   0%|          | 0.00/480 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/331M [00:00<?, ?B/s]

Some weights of the model checkpoint at distilbert/distilroberta-base were not used when initializing RobertaForMaskedLM: ['roberta.pooler.dense.bias', 'roberta.pooler.dense.weight']
- This IS expected if you are initializing RobertaForMaskedLM from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing RobertaForMaskedLM from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).


tokenizer_config.json:   0%|          | 0.00/25.0 [00:00<?, ?B/s]

vocab.json: 0.00B [00:00, ?B/s]

merges.txt: 0.00B [00:00, ?B/s]

tokenizer.json: 0.00B [00:00, ?B/s]

Device set to use cuda:0


[{'score': 0.0935290977358818,
  'token': 774,
  'token_str': ' role',
  'sequence': 'This couse will teach you all about role models.'},
 {'score': 0.05528588965535164,
  'token': 3477,
  'token_str': ' animal',
  'sequence': 'This couse will teach you all about animal models.'}]

### 1.5 NER Pipeline
The NER pipeline identifies entities such as persons, organizations or locations in a sentence.

In [ ]:
ner = pipeline("ner", grouped_entities=True)
ner("My name is John Doe and I study at Stanford in USA.")

No model was supplied, defaulted to dbmdz/bert-large-cased-finetuned-conll03-english and revision 4c53496 (https://huggingface.co/dbmdz/bert-large-cased-finetuned-conll03-english).
Using a pipeline without specifying a model name and revision in production is not recommended.


config.json:   0%|          | 0.00/998 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/1.33G [00:00<?, ?B/s]

Some weights of the model checkpoint at dbmdz/bert-large-cased-finetuned-conll03-english were not used when initializing BertForTokenClassification: ['bert.pooler.dense.bias', 'bert.pooler.dense.weight']
- This IS expected if you are initializing BertForTokenClassification from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing BertForTokenClassification from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).


tokenizer_config.json:   0%|          | 0.00/60.0 [00:00<?, ?B/s]

vocab.txt: 0.00B [00:00, ?B/s]

Device set to use cuda:0
/usr/local/lib/python3.12/dist-packages/transformers/pipelines/token_classification.py:186: UserWarning: `grouped_entities` is deprecated and will be removed in version v5.0.0, defaulted to `aggregation_strategy="AggregationStrategy.SIMPLE"` instead.
  warnings.warn(


[{'entity_group': 'PER',
  'score': np.float32(0.9977967),
  'word': 'John Doe',
  'start': 11,
  'end': 19},
 {'entity_group': 'ORG',
  'score': np.float32(0.9635525),
  'word': 'Stanford',
  'start': 35,
  'end': 43},
 {'entity_group': 'LOC',
  'score': np.float32(0.99960965),
  'word': 'USA',
  'start': 47,
  'end': 50}]

## 1.6 Question Answering

In [ ]:
qa = pipeline("question-answering")
qa(
    question="Where do I work?",
    context="My name is Sagar and I work at Facebook in California"
)

No model was supplied, defaulted to distilbert/distilbert-base-cased-distilled-squad and revision 564e9b5 (https://huggingface.co/distilbert/distilbert-base-cased-distilled-squad).
Using a pipeline without specifying a model name and revision in production is not recommended.


config.json:   0%|          | 0.00/473 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/261M [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/49.0 [00:00<?, ?B/s]

vocab.txt: 0.00B [00:00, ?B/s]

tokenizer.json: 0.00B [00:00, ?B/s]

Fetching 0 files: 0it [00:00, ?it/s]

Fetching 1 files:   0%|          | 0/1 [00:00<?, ?it/s]

Fetching 0 files: 0it [00:00, ?it/s]

Device set to use cuda:0


{'score': 0.6411417722702026, 'start': 31, 'end': 39, 'answer': 'Facebook'}

## 1.7 Summarization

In [ ]:
summarizer = pipeline("summarization")
summarizer("""
    The quiet rustle of leaves in the early morning breeze is a gentle reminder that the world doesn’t need to shout to be beautiful.
    Somewhere between the dew-soaked blades of grass and the slow stretch of sunlight across the horizon,
    there's a kind of peace that cities can’t replicate. Birds don’t hurry their songs,
    and the trees aren’t in competition to grow the fastest — everything just exists, calmly and without apology.
""")

No model was supplied, defaulted to sshleifer/distilbart-cnn-12-6 and revision a4f8f3e (https://huggingface.co/sshleifer/distilbart-cnn-12-6).
Using a pipeline without specifying a model name and revision in production is not recommended.


config.json: 0.00B [00:00, ?B/s]

pytorch_model.bin:   0%|          | 0.00/1.22G [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/1.22G [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/26.0 [00:00<?, ?B/s]

vocab.json: 0.00B [00:00, ?B/s]

merges.txt: 0.00B [00:00, ?B/s]

Device set to use cuda:0
Your max_length is set to 142, but your input_length is only 119. Since this is a summarization task, where outputs shorter than the input are typically wanted, you might consider decreasing max_length manually, e.g. summarizer('...', max_length=59)


[{'summary_text': " The quiet rustle of leaves in the early morning breeze is a gentle reminder that the world doesn't need to shout to be beautiful . Somewhere between the dew-soaked blades of grass and the slow stretch of sunlight across the horizon, there's a kind of peace that cities can't replicate ."}]

## Advanced Options
You can control the behavior of pipelines using various arguments:

For text-generation (e.g., GPT-2):

In [ ]:
generator = pipeline("text-generation", model="gpt2")

output = generator(
    "Once upon a time",
    max_length=50,
    num_return_sequences=2,
    do_sample=True,
    temperature=0.7,
    top_k=50,
    top_p=0.95
)

for i, text in enumerate(output):
    print(f"Result {i+1}:\n{text['generated_text']}\n")


config.json:   0%|          | 0.00/665 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/548M [00:00<?, ?B/s]

generation_config.json:   0%|          | 0.00/124 [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/26.0 [00:00<?, ?B/s]

vocab.json:   0%|          | 0.00/1.04M [00:00<?, ?B/s]

merges.txt:   0%|          | 0.00/456k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/1.36M [00:00<?, ?B/s]

Device set to use cuda:0
Truncation was not explicitly activated but `max_length` is provided a specific value, please use `truncation=True` to explicitly truncate examples to max length. Defaulting to 'longest_first' truncation strategy. If you encode pairs of sequences (GLUE-style) with the tokenizer you can select this strategy more precisely by providing a specific strategy to `truncation`.
Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.
Both `max_new_tokens` (=256) and `max_length`(=50) seem to have been set. `max_new_tokens` will take precedence. Please refer to the documentation for more information. (https://huggingface.co/docs/transformers/main/en/main_classes/text_generation)


Result 1:
Once upon a time, you were able to create a great work of art. But that work is now gone. And the rest is history.

The truth is that the art world is in such a state of disrepair, that many artists now simply cannot do their work. They are forced to do the work of their dreams.

And it is this kind of disrepair that is causing the art world to be so sad.

This is why I am doing this. Because the art world is in such a state of disrepair, that many artists now simply cannot do their work. They are forced to do the work of their dreams.

This is why I am doing this. Because the art world is in such a state of disrepair, that many artists now simply cannot do their work. They are forced to do the work of their dreams.

And that is why I am doing this. Because the art world is in such a state of disrepair, that many artists now simply cannot do their work. They are forced to do the work of their dreams.

It is just as well that artists can't afford to leave their home to create,

#### Internals (Behind the Scenes)
- When you run pipeline(task):
    1. Task type maps to a pipeline class (e.g., TextClassificationPipeline)
    2. Loads the default model/tokenizer if not specified
    3. Handles batching, truncation, padding, and special tokens
    4. Runs inference
    5. Applies task-specific postprocessing (e.g., softmax, argmax, span extraction)

## Translation

In [ ]:
from transformers import pipeline

# Create a translation pipeline (English → French)
translator = pipeline("translation_en_to_fr")

# Text to translate
text = "Hello! How are you today?"

# Translate the text
result = translator(text)

# Print the translated output
print("Translated:", result[0]['translation_text'])


No model was supplied, defaulted to google-t5/t5-base and revision a9723ea (https://huggingface.co/google-t5/t5-base).
Using a pipeline without specifying a model name and revision in production is not recommended.


config.json: 0.00B [00:00, ?B/s]

model.safetensors:   0%|          | 0.00/892M [00:00<?, ?B/s]

generation_config.json:   0%|          | 0.00/147 [00:00<?, ?B/s]

spiece.model:   0%|          | 0.00/792k [00:00<?, ?B/s]

tokenizer.json: 0.00B [00:00, ?B/s]

Device set to use cuda:0


Translated: Bonjour, comment vous êtes-vous aujourd'hui?


## 1.8 Zero-Shot-Classification

With zero-shot classification, you can give the model:

A text (e.g. "I love watching soccer.")

A list of candidate labels (e.g. ["sports", "politics", "technology"])

...and the model will predict which label(s) best fit the text, even though it hasn’t been specifically trained on those categories.

In [ ]:
from transformers import pipeline

# Create a zero-shot classifier
classifier = pipeline("zero-shot-classification")

# Text to classify
text = "The new smartphone has an amazing camera and long battery life."

# Candidate labels (can be anything!)
labels = ["technology", "sports", "food", "travel", "entertainment"]

result = classifier(text, labels)

print(f"Text: '{text}'\n")
print("Classification Results:")
for label, score in zip(result['labels'], result['scores']):
    print(f"  - {label}: {score:.2%}")

No model was supplied, defaulted to facebook/bart-large-mnli and revision d7645e1 (https://huggingface.co/facebook/bart-large-mnli).
Using a pipeline without specifying a model name and revision in production is not recommended.


config.json: 0.00B [00:00, ?B/s]

model.safetensors:   0%|          | 0.00/1.63G [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/26.0 [00:00<?, ?B/s]

vocab.json: 0.00B [00:00, ?B/s]

merges.txt: 0.00B [00:00, ?B/s]

tokenizer.json: 0.00B [00:00, ?B/s]

Device set to use cuda:0


Text: 'The new smartphone has an amazing camera and long battery life.'

Classification Results:
  - technology: 93.59%
  - entertainment: 3.29%
  - travel: 1.61%
  - sports: 1.27%
  - food: 0.24%


## 1.9 Feature Extraction

In [ ]:
from transformers import pipeline
features = pipeline("feature-extraction")
vector = features("I love AI")[0]

len(vector)

No model was supplied, defaulted to distilbert/distilbert-base-cased and revision 6ea8117 (https://huggingface.co/distilbert/distilbert-base-cased).
Using a pipeline without specifying a model name and revision in production is not recommended.


config.json:   0%|          | 0.00/465 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/263M [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/49.0 [00:00<?, ?B/s]

vocab.txt: 0.00B [00:00, ?B/s]

tokenizer.json: 0.00B [00:00, ?B/s]

Device set to use cuda:0


5

In [ ]:
feature_extractor = pipeline("feature-extraction")
text = "Hello world"
embeddings = feature_extractor(text)
embeddings

No model was supplied, defaulted to distilbert/distilbert-base-cased and revision 6ea8117 (https://huggingface.co/distilbert/distilbert-base-cased).
Using a pipeline without specifying a model name and revision in production is not recommended.
Device set to use cuda:0


[[[0.3180203139781952,
   0.09429965168237686,
   0.006127363070845604,
   -0.34622055292129517,
   -0.4438483417034149,
   -0.06749323010444641,
   0.29249483346939087,
   -0.017795803025364876,
   0.08857250958681107,
   -1.193962812423706,
   -0.35154882073402405,
   0.16215382516384125,
   -0.07816784083843231,
   -0.059919487684965134,
   -0.6221104860305786,
   0.18003025650978088,
   0.09457312524318695,
   -0.03020242601633072,
   -0.2581993639469147,
   -0.1877536177635193,
   0.06186233460903168,
   -0.23031161725521088,
   0.5440695881843567,
   -0.19957034289836884,
   0.22743995487689972,
   0.0029055557679384947,
   0.31153014302253723,
   0.173297718167305,
   -0.2075628638267517,
   0.3399984538555145,
   0.10412824898958206,
   0.23860043287277222,
   -0.07771787792444229,
   0.1253519058227539,
   -0.11217060685157776,
   0.0027829348109662533,
   0.0037186460103839636,
   -0.2642000913619995,
   -0.008924736641347408,
   -0.16470806300640106,
   -0.49579060077667236,

In [ ]:
from transformers import pipeline
import numpy as np

# Create a feature extractor
feature_extractor = pipeline("feature-extraction")

# Texts to compare
texts = [
    "I love machine learning",
    "AI is fascinating",
    "The weather is nice today"
]

# Get embeddings (numerical representations)
embeddings = []
for text in texts:
    embedding = np.array(feature_extractor(text)).mean(axis=1).squeeze()
    embeddings.append(embedding)
    print(f"'{text}' -> Vector shape: {embedding.shape}")

# Show similarity between first two texts
similarity = np.dot(embeddings[0], embeddings[1]) / (np.linalg.norm(embeddings[0]) * np.linalg.norm(embeddings[1]))
print(f"\n🔗 Similarity between '{texts[0]}' and '{texts[1]}': {similarity:.2%}")

No model was supplied, defaulted to distilbert/distilbert-base-cased and revision 6ea8117 (https://huggingface.co/distilbert/distilbert-base-cased).
Using a pipeline without specifying a model name and revision in production is not recommended.
Device set to use cuda:0


'I love machine learning' -> Vector shape: (768,)
'AI is fascinating' -> Vector shape: (768,)
'The weather is nice today' -> Vector shape: (768,)

🔗 Similarity between 'I love machine learning' and 'AI is fascinating': 89.75%


##  What’s Happening Here?

### 1. **`pipeline("feature-extraction")`**

* Loads a default model (usually `bert-base-uncased`) that turns text into **hidden states** (embeddings) — one vector per token.
* We’re averaging the token vectors to get a **sentence-level vector**.

---

### 2. **Averaging the Embeddings**

```python
embedding = np.array(feature_extractor(text)).mean(axis=1).squeeze()
```

* `feature_extractor(text)` returns a shape: `[1, sequence_length, hidden_size]`

  * For BERT: typically `[1, ~10, 768]`
* `.mean(axis=1)` → Averages across tokens (sequence length) → `[1, 768]`
* `.squeeze()` → Removes the extra dimension → `[768]`

So each sentence becomes a single **768-dimensional vector**.

---

### 3. **Similarity Score**

```python
similarity = cosine_similarity(embedding1, embedding2)
```

This computes how similar two texts are in vector space:

* `1.0` = very similar
* `0.0` = orthogonal (unrelated)
* `-1.0` = opposite meaning (rare in practice with BERT)

---

##  Sample Output (what we'd expect):

```text
'I love machine learning' -> Vector shape: (768,)
'AI is fascinating' -> Vector shape: (768,)
'The weather is nice today' -> Vector shape: (768,)

🔗 Similarity between 'I love machine learning' and 'AI is fascinating': 86.45%
```

(our score may vary a bit due to model internals.)

---

##  Improvements / Customization Ideas

### Use a Model You Choose (e.g., sentence transformers)

```python
feature_extractor = pipeline(
    "feature-extraction",
    model="sentence-transformers/all-MiniLM-L6-v2"
)
```

This model gives **better sentence-level embeddings**, as it’s trained specifically for that.



## Image Captioner

In [ ]:
from transformers import pipeline

captioner = pipeline(model="ydshieh/vit-gpt2-coco-en")
captioner("https://huggingface.co/datasets/Narsil/image_dummy/raw/main/parrots.png")

config.json: 0.00B [00:00, ?B/s]

/usr/local/lib/python3.12/dist-packages/transformers/models/auto/modeling_auto.py:2242: FutureWarning: The class `AutoModelForVision2Seq` is deprecated and will be removed in v5.0. Please use `AutoModelForImageTextToText` instead.
  warnings.warn(


pytorch_model.bin:   0%|          | 0.00/982M [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/982M [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/236 [00:00<?, ?B/s]

vocab.json: 0.00B [00:00, ?B/s]

merges.txt: 0.00B [00:00, ?B/s]

tokenizer.json: 0.00B [00:00, ?B/s]

special_tokens_map.json:   0%|          | 0.00/120 [00:00<?, ?B/s]

Fetching 1 files:   0%|          | 0/1 [00:00<?, ?it/s]

preprocessor_config.json:   0%|          | 0.00/211 [00:00<?, ?B/s]

Using a slow image processor as `use_fast` is unset and a slow processor was saved with this model. `use_fast=True` will be the default behavior in v4.52, even if the model was saved with a slow processor. This will result in minor differences in outputs. You'll still be able to use a slow processor with `use_fast=False`.
Device set to use cuda:0
The attention mask is not set and cannot be inferred from input because pad token is same as eos token. As a consequence, you may observe unexpected behavior. Please pass your input's `attention_mask` to obtain reliable results.
We strongly recommend passing in an `attention_mask` since your input_ids may be padded. See https://huggingface.co/docs/transformers/troubleshooting#incorrect-output-when-padding-tokens-arent-masked.
You may ignore this warning if your `pad_token_id` (50256) is identical to the `bos_token_id` (50256), `eos_token_id` (50256), or the `sep_token_id` (None), and your input is not padded.


[{'generated_text': 'two birds are standing next to each other '}]

## Text to Audio

In [ ]:
!pip install  gtts

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 98.2/98.2 kB 9.4 MB/s eta 0:00:00
  Attempting uninstall: click
    Found existing installation: click 8.2.1
    Uninstalling click-8.2.1:
      Successfully uninstalled click-8.2.1


In [ ]:
# First install: pip install gtts
from gtts import gTTS
import os

def text_to_speech_google():
    texts = [
        "Hello! Welcome to the world of AI and text to speech technology.",
        "This is amazing! I can convert any text into natural sounding speech.",
        "The possibilities are endless with artificial intelligence."
    ]

    for i, text in enumerate(texts):
        print(f"🗣️ Converting: '{text}'")

        # Create Google TTS object
        tts = gTTS(text=text, lang='en', slow=False)

        # Save to file
        output_file = f"speech_output_{i}.mp3"
        tts.save(output_file)
        print(f"💾 Saved as: {output_file}")
        print(f"🔊 Play the file: {output_file}")

# Run it
text_to_speech_google()

🗣️ Converting: 'Hello! Welcome to the world of AI and text to speech technology.'
💾 Saved as: speech_output_0.mp3
🔊 Play the file: speech_output_0.mp3
🗣️ Converting: 'This is amazing! I can convert any text into natural sounding speech.'
💾 Saved as: speech_output_1.mp3
🔊 Play the file: speech_output_1.mp3
🗣️ Converting: 'The possibilities are endless with artificial intelligence.'
💾 Saved as: speech_output_2.mp3
🔊 Play the file: speech_output_2.mp3


In [ ]:
from IPython.display import Audio, display

file_0 = "/content/speech_output_0.mp3"
file_1 = "/content/speech_output_1.mp3"
file_2 = "/content/speech_output_2.mp3"

# After saving the file
display(Audio(file_0))
display(Audio(file_1))
display(Audio(file_2))


# Gradio Application

No need to dive into the code, just use the app

In [ ]:
# Install required packages
!pip install transformers gradio torch sentencepiece

import gradio as gr
from transformers import pipeline

print("🚀 Setting up AI pipelines...")

# Initialize all pipelines with specific models for better performance
pipelines = {
    "sentiment": pipeline("sentiment-analysis", model="distilbert-base-uncased-finetuned-sst-2-english"),
    "text_generation": pipeline("text-generation", model="gpt2", max_length=100),
    "qa": pipeline("question-answering", model="distilbert-base-cased-distilled-squad"),
    "translation": pipeline("translation_en_to_fr", model="t5-small"),
    "summarization": pipeline("summarization", model="facebook/bart-large-cnn"),
    "ner": pipeline("ner", aggregation_strategy="simple", model="dbmdz/bert-large-cased-finetuned-conll03-english"),
    "zero_shot": pipeline("zero-shot-classification", model="facebook/bart-large-mnli")
}

print("✅ All pipelines loaded successfully!")

def ai_playground(task, text, context, labels):
    try:
        if task == "sentiment":
            result = pipelines["sentiment"](text)[0]
            emotion = "😊" if result['label'] == 'POSITIVE' else '😞'
            return f"{emotion} {result['label']} ({result['score']:.2%} confidence)"

        elif task == "text_generation":
            result = pipelines["text_generation"](text, max_length=100, do_sample=True)[0]['generated_text']
            return f"📝 {result}"

        elif task == "qa":
            if not context.strip():
                return "❌ Please provide context for question answering"
            result = pipelines["qa"](question=text, context=context)
            return f"✅ Answer: {result['answer']}\n🎯 Confidence: {result['score']:.2%}"

        elif task == "translation":
            result = pipelines["translation"](text)[0]['translation_text']
            return f"🇫🇷 {result}"

        elif task == "summarization":
            if len(text) < 50:
                return "❌ Text is too short for summarization. Please provide longer text."
            result = pipelines["summarization"](text, max_length=80, min_length=30, do_sample=False)[0]['summary_text']
            return f"📌 Summary: {result}"

        elif task == "ner":
            entities = pipelines["ner"](text)
            if not entities:
                return "🔍 No named entities found"
            result = "\n".join([f"• {entity['word']} → {entity['entity_group']} ({entity['score']:.2%})" for entity in entities[:10]])
            return f"🔍 Found Entities:\n{result}"

        elif task == "zero_shot":
            if not labels.strip():
                return "❌ Please provide labels for zero-shot classification"
            label_list = [label.strip() for label in labels.split(",") if label.strip()]
            result = pipelines["zero_shot"](text, label_list, multi_label=False)
            formatted_results = "\n".join([f"• {label}: {score:.2%}" for label, score in zip(result['labels'][:5], result['scores'][:5])])
            return f"🏷️ Classification:\n{formatted_results}"

        else:
            return "Please provide required inputs"

    except Exception as e:
        return f"❌ Error: {str(e)}"

# Create the interface with better layout
with gr.Blocks(theme=gr.themes.Soft(), title="🎪 AI Playground") as demo:
    gr.Markdown("# 🎪 AI Playground - Try Everything!")
    gr.Markdown("Perfect for beginners! Choose a task and see AI in action. Built for Google Colab! 🚀")

    with gr.Row():
        with gr.Column(scale=1):
            task_dropdown = gr.Dropdown(
                choices=["sentiment", "text_generation", "qa", "translation", "summarization", "ner", "zero_shot"],
                label="🎯 Choose AI Task",
                value="sentiment",
                info="Select what you want to do"
            )

            input_text = gr.Textbox(
                lines=3,
                placeholder="Enter your text here...",
                label="📝 Input Text",
                value="I love learning about artificial intelligence!"
            )

            context_text = gr.Textbox(
                lines=3,
                placeholder="For QA: Enter context here...\nExample: Hugging Face provides AI tools and models. The company was founded in 2016.",
                label="📚 Context (for QA only)",
                visible=False
            )

            labels_text = gr.Textbox(
                lines=2,
                placeholder="For Zero-shot: Enter comma-separated labels\nExample: positive, negative, neutral",
                label="🏷️ Labels (for Zero-shot only)",
                visible=False
            )

            submit_btn = gr.Button("✨ Run AI Magic!", variant="primary")

        with gr.Column(scale=2):
            output_text = gr.Textbox(
                lines=8,
                label="🎉 AI Output",
                interactive=False,
                show_copy_button=True
            )

    # Examples for each task
    gr.Markdown("## 💡 Try These Examples:")

    examples = [
        ["sentiment", "I'm so excited to learn about AI!", "", ""],
        ["text_generation", "Once upon a time in a magical kingdom,", "", ""],
        ["translation", "Hello, how are you today?", "", ""],
        ["summarization", "Artificial intelligence is transforming many industries. Machine learning allows computers to learn from data. Deep learning uses neural networks for complex tasks. AI is used in healthcare, finance, and transportation.", "", ""],
        ["ner", "Apple was founded by Steve Jobs in Cupertino, California in 1976.", "", ""],
        ["zero_shot", "The new smartphone has amazing battery life and camera quality.", "", "technology, review, complaint, inquiry"],
        ["qa", "What does Hugging Face provide?", "Hugging Face is a company that provides AI tools and models for natural language processing. Their library makes it easy to use state-of-the-art models.", ""]
    ]

    gr.Examples(
        examples=examples,
        inputs=[task_dropdown, input_text, context_text, labels_text],
        outputs=output_text,
        fn=ai_playground,
        cache_examples=False,
        label="Click any example below to try it!"
    )

    # Show/hide context and labels based on task selection
    def toggle_inputs(task):
        context_visible = (task == "qa")
        labels_visible = (task == "zero_shot")
        return [
            gr.Textbox(visible=context_visible),
            gr.Textbox(visible=labels_visible)
        ]

    task_dropdown.change(
        fn=toggle_inputs,
        inputs=task_dropdown,
        outputs=[context_text, labels_text]
    )

    # Connect the submit button
    submit_btn.click(
        fn=ai_playground,
        inputs=[task_dropdown, input_text, context_text, labels_text],
        outputs=output_text
    )

    gr.Markdown("---")
    gr.Markdown("### 🎓 Learning Tips:")
    gr.Markdown("""
    - **Sentiment Analysis**: Detects positive/negative emotions in text
    - **Text Generation**: Creates new text based on your prompt
    - **Translation**: Translates English to French
    - **Summarization**: Shortens long text while keeping key points
    - **Named Entity Recognition**: Finds people, places, organizations in text
    - **Zero-Shot Classification**: Classifies text into custom categories
    - **Question Answering**: Answers questions based on provided context
    """)

print("🎉 AI Playground is ready! Launching interface...")
demo.launch(share=True, debug=True)

🚀 Setting up AI pipelines...


config.json:   0%|          | 0.00/629 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/268M [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/48.0 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/232k [00:00<?, ?B/s]

Device set to use cuda:0
Device set to use cuda:0


config.json:   0%|          | 0.00/473 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/261M [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/49.0 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/213k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/436k [00:00<?, ?B/s]

Fetching 0 files: 0it [00:00, ?it/s]

Fetching 1 files:   0%|          | 0/1 [00:00<?, ?it/s]

Fetching 0 files: 0it [00:00, ?it/s]

Device set to use cuda:0


config.json:   0%|          | 0.00/1.21k [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/242M [00:00<?, ?B/s]

generation_config.json:   0%|          | 0.00/147 [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/2.32k [00:00<?, ?B/s]

spiece.model:   0%|          | 0.00/792k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/1.39M [00:00<?, ?B/s]

Device set to use cuda:0


config.json: 0.00B [00:00, ?B/s]

model.safetensors:   0%|          | 0.00/1.63G [00:00<?, ?B/s]

generation_config.json:   0%|          | 0.00/363 [00:00<?, ?B/s]

vocab.json: 0.00B [00:00, ?B/s]

merges.txt: 0.00B [00:00, ?B/s]

tokenizer.json: 0.00B [00:00, ?B/s]

Device set to use cuda:0
Some weights of the model checkpoint at dbmdz/bert-large-cased-finetuned-conll03-english were not used when initializing BertForTokenClassification: ['bert.pooler.dense.bias', 'bert.pooler.dense.weight']
- This IS expected if you are initializing BertForTokenClassification from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing BertForTokenClassification from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).
Device set to use cuda:0
Device set to use cuda:0


✅ All pipelines loaded successfully!
🎉 AI Playground is ready! Launching interface...
Colab notebook detected. This cell will run indefinitely so that you can see errors and logs. To turn off, set debug=False in launch().
* Running on public URL: https://d4756dc24c96ac0c29.gradio.live

This share link expires in 1 week. For free permanent hosting and GPU upgrades, run `gradio deploy` from the terminal in the working directory to deploy to Hugging Face Spaces (https://huggingface.co/spaces)


Truncation was not explicitly activated but `max_length` is provided a specific value, please use `truncation=True` to explicitly truncate examples to max length. Defaulting to 'longest_first' truncation strategy. If you encode pairs of sequences (GLUE-style) with the tokenizer you can select this strategy more precisely by providing a specific strategy to `truncation`.
Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.


Keyboard interruption in main thread... closing server.
Killing tunnel 127.0.0.1:7860 <> https://d4756dc24c96ac0c29.gradio.live
